### PR TITLE
refactor(react/hooks): unnecesary hooks and reusability

### PR DESCRIPTION
**Pull request by andresin87**
- Removes useSetState (unnecesary)
- Refactor of useMount (removes unused arguments)
- Refactor and use useMount on useControlledState
- Refactor useCopyToClipboard (pase useSetState usage to useState)

### CODE REVIEW

File: components/react/hooks/src/useCopyToClipboard/index.js [hunk 1]
```diff
@@ -1,15 +1,12 @@
...
 import useMountedState from '../useMountedState'
-import useSetState from '../useSetState'
 
 export default function useCopyToClipboard() {
   const isMounted = useMountedState()
-  const [state, setState] = useSetState({
-    value: undefined,
-    error: undefined
-  })
+  const [value, setValue] = useState()
```
Comment (jcger) [inline components/react/hooks/src/useCopyToClipboard/index.js:8]: as you are setting these values always as one, what about using something like `const [{value, error}, setState]` so you can `setState({value, error})` afterwards without triggering 2 renders?

Comment (andresin87) [reply components/react/hooks/src/useCopyToClipboard/index.js]: done

File: components/react/hooks/src/useMountedState/index.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import {useCallback, useEffect, useRef} from 'react'
 
-export default function useMountedState(value, delay) {
+export default function useMountedState() {
```
Comment (tonyghiani) [inline components/react/hooks/src/useMountedState/index.js:3]: I have doubts about this hook, why not returning this one as a boolean directly more than a getter function?

File: components/react/hooks/src/useControlledState/index.js [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
     controlledValue === undefined ? defaultValue : controlledValue
   )
   const [value, setValue] = useState(initialValue)
-  useEffect(() => {
-    if (isFirst.current) {
-      isFirst.current = false
-    } else {
+  const isMounted = useMountedState()
+  useLayoutEffect(() => {
+    if (isMounted()) {
```
Comment (rmoralp) [inline components/react/hooks/src/useControlledState/index.js:14]: I think that it changes behaviour. In changed code isFirst was set to true, then moved to false after "mount" and then `setValue`
Isn't it?
```suggestion
    if (isMounted() === false) {
```

File: components/react/hooks/src/useControlledState/index.js [hunk 4]
```diff
@@ -1,21 +1,20 @@
...
   const [initialValue] = useState(
     controlledValue === undefined ? defaultValue : controlledValue
   )
   const [value, setValue] = useState(initialValue)
-  useEffect(() => {
-    if (isFirst.current) {
-      isFirst.current = false
-    } else {
+  const isMounted = useMountedState()
+  useLayoutEffect(() => {
```
[the remark this answers is not in the document]
Comment (andresin87) [reply components/react/hooks/src/useControlledState/index.js:13]: isMount uses the useEffect on it and use the useRef for the mounted status acknowledge. In order not to depend on the useEffect declaration position preventing posible bugs we might use This useLayoutEffect.

[@kentcdodds](https://kentcdodds.com/blog/useeffect-vs-uselayouteffect) (last example –One special case
–)

File: components/react/hooks/src/useControlledState/index.js (hunk 3 above)
Comment (andresin87) [reply components/react/hooks/src/useControlledState/index.js:14]: Its a semantic difference. isFirst means isNotMounted

File: components/react/hooks/src/useMountedState/index.js (hunk 2 above)
Comment (andresin87) [reply components/react/hooks/src/useMountedState/index.js:3]: It is using the reference for the behavior and no-useEffect should depend on the state of the hooks inner value, so we only should return a memoized callback targetting to the value (no other component observers will be refreshed when its inner status is updated)

take a look to the [https://github.com/streamich/react-use/blob/master/src/useMountedState.ts](https://github.com/streamich/react-use/blob/master/src/useMountedState.ts) code example

Comment (tonyghiani) [reply components/react/hooks/src/useMountedState/index.js:3]: Sure i understand, what i mean is that even returning directly the boolean value and not a getter it shouldn't notify and rerender any component observing this value

**Review by tonyghiani — APPROVED** [review]

**Review by jcger — APPROVED** [review]

**Pull request merged**